### PR TITLE
[FeG][Orc8r] Add charging characteristics filter for Virtual APN

### LINKED ISSUE #4162: [VirtualAPN] APN replacement based on charging characteristics
Target for 1.4

### Describe the Issue

Our current VirtualAPN config is the following:
```
type VirtualApnRule struct {

	// Regex to match APN values
	ApnFilter string `json:"apn_filter,omitempty"`

	// New APN to overwrite when filter matches
	ApnOverwrite string `json:"apn_overwrite,omitempty"`
}
```
where SessionProxy replaces any APNs that match the APNFilter (RegEx) with the ApnOverwrite value.
This is not sufficient as we want to be able to also case on the charging characteristics as well.

The goal of this issue is to extend this configuration to include a `ChargingCharacteristicsFilter` (RegEx) to also case on the incoming session context.
The new logic will be: any session with APN that satisfies `ApnFilter` and Charging Characteristics that satisfies `ChargingCharacteristicsFilter` will have its APN replaced with `ApnOverwrite`.

NEW
```
type VirtualApnRule struct {

	// Regex to match APN values
	ApnFilter string `json:"apn_filter,omitempty"`

	// Regex to match Charging Characteristics values
	ChargingCharacteristicsFilter string `json:"charging_characteristics_filter,omitempty"`

	// New APN to overwrite when filter matches
	ApnOverwrite string `json:"apn_overwrite,omitempty"`
}
```

### PR DESCRIPTION
**Pull request by themarwhal**
Signed-off-by: Marie Bremner <marwhal@fb.com>

## Summary

Our current VirtualAPN config is the following:
```
type VirtualApnRule struct {

	// Regex to match APN values
	ApnFilter string `json:"apn_filter,omitempty"`

	// New APN to overwrite when filter matches
	ApnOverwrite string `json:"apn_overwrite,omitempty"`
}
```
where SessionProxy replaces any APNs that match the APNFilter (RegEx) with the ApnOverwrite value.
This is not sufficient as we want to be able to also case on the charging characteristics as well.

The goal of this issue is to extend this configuration to include a `ChargingCharacteristicsFilter` (RegEx) to also case on the incoming session context.
The new logic will be: any session with APN that satisfies `ApnFilter` and Charging Characteristics that satisfies `ChargingCharacteristicsFilter` will have its APN replaced with `ApnOverwrite`.

NEW
```
type VirtualApnRule struct {

	// Regex to match APN values
	ApnFilter string `json:"apn_filter,omitempty"`

	// Regex to match Charging Characteristics values
	ChargingCharacteristicsFilter string `json:"charging_characteristics_filter,omitempty"`

	// New APN to overwrite when filter matches
	ApnOverwrite string `json:"apn_overwrite,omitempty"`
}
```
## Test Plan
FeG precommit, Orc8r precommit
Local Orc8r API testing
Some hacky testing with the CWF integ test setup to test CCR creations from a static gateway.mconfig

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: feg/gateway/services/session_proxy/credit_control/virtual_apn_rules.go [hunk 1]
```diff
@@ -15,43 +15,55 @@ package credit_control
...
+		apnRule := &VirtualApnRule{}
+		err := apnRule.FromMconfig(pVirtualApnCfg)
 		if err != nil {
-			log.Printf("%s Managed Gx Virtual APN Rule Config Load Error: %v", virtualApnlCfg.GetApnFilter(), err)
+			glog.Errorf("%s Managed Virtual APN Rule Config Load Error: %v", pVirtualApnCfg.GetApnFilter(), err)
 			continue
 		}
 		virtualApnConfigs = append(virtualApnConfigs, apnRule)
-		log.Printf("Virtual APN Rule Activated filter: %s, Overwrite: %s", virtualApnlCfg.GetApnFilter(), virtualApnlCfg.GetApnOverwrite())
+		glog.Infof("Virtual APN Rule Activated filter: %s, Overwrite: %s", pVirtualApnCfg.GetApnFilter(), pVirtualApnCfg.GetApnOverwrite())
```
Comment (uri200) [inline feg/gateway/services/session_proxy/credit_control/virtual_apn_rules.go]: not sure if `GenerateVirtualApnRules` is triggered per each request or just at init. But if this is per request, then we should be setting this as `V(2)`

**Review by uri200 — APPROVED** [review]

File: feg/gateway/services/session_proxy/credit_control/virtual_apn_rules.go (hunk 1 above)
Comment (themarwhal) [reply feg/gateway/services/session_proxy/credit_control/virtual_apn_rules.go]: This is done everytime the service starts. So Info is probably OK. :)

File: feg/protos/mconfig/mconfigs.proto [hunk 2]
```diff
@@ -56,8 +56,12 @@ message S6aConfig {
 }
 
 message VirtualApnRule {
+    // Regex to match APN values
```
Comment (hcgatewood) [inline feg/protos/mconfig/mconfigs.proto]: Here and elsewhere -- prefer golang-style doc comments, i.e. start the comment with `// apn_filter is a regex ...`

File: lte/protos/session_manager.proto [hunk 3]
```diff
@@ -420,6 +420,7 @@ message CreditUsageUpdate {
   TgppContext tgpp_ctx = 15;
   CommonSessionContext common_context = 16;
   RatSpecificContext rat_specific_context = 17;
+  string charging_characteristics = 18;
```
Comment (hcgatewood) [inline lte/protos/session_manager.proto:423]: Not for this PR but: the lack of doc comments in this proto file is a bit sad 😥

File: feg/gateway/services/session_proxy/credit_control/gx/gx_client.go [hunk 4]
```diff
@@ -430,15 +430,15 @@ func (gxClient *GxClient) getEventTriggerAVP(eventTrigger EventTrigger) *diam.AV
 	return diam.NewAVP(avp.EventTrigger, avp.Mbit|avp.Vbit, diameter.Vendor3GPP, datatype.Enumerated(eventTrigger))
 }
 
-// getAPNfromConfig returns a new apn value to overwrite the one in the request based on list of regex definied in Gx config.
+// getAPNFromConfig returns a new apn value to overwrite the one in the request based on list of regex definied in Gx config.
```
Comment (hcgatewood) [inline feg/gateway/services/session_proxy/credit_control/gx/gx_client.go]: Golang convention (in Magma codebase and in stdlib) is to wrap comments to 79 chars

**Review by hcgatewood — APPROVED** [review]
Cloud changes lgtm

**Pull request merged**